src/components/Asset/AssetActions/ButtonBuy/index.test.tsx

```diff
@@ -0,0 +1,146 @@
+import React from 'react'
+import { render, screen } from '@testing-library/react'
+import ButtonBuy, { ButtonBuyProps } from './'
+
+const downloadProps: ButtonBuyProps = {
+  action: 'download',
+  disabled: false,
+  hasPreviousOrder: false,
+  hasDatatoken: false,
+  btSymbol: 'btSymbol',
+  dtSymbol: 'dtSymbol',
+  dtBalance: '100000000000',
+  assetTimeout: '1 day',
+  assetType: 'Dataset',
+  stepText: 'TEST',
+  priceType: 'fixed',
+  isConsumable: true,
+  isBalanceSufficient: true,
+  consumableFeedback: 'TEST: consumableFeedback'
+}
+
+const computeProps: ButtonBuyProps = {
+  action: 'compute',
+  disabled: false,
+  hasPreviousOrder: false,
+  hasDatatoken: true,
+  btSymbol: 'btSymbol',
+  dtSymbol: 'dtSymbol',
+  dtBalance: '100000000000',
+  assetTimeout: '1 day',
+  assetType: 'algorithm',
+  hasPreviousOrderSelectedComputeAsset: false,
+  hasDatatokenSelectedComputeAsset: true,
+  dtSymbolSelectedComputeAsset: 'dtSymbol',
+  dtBalanceSelectedComputeAsset: 'dtBalance',
+  selectedComputeAssetType: 'selectedComputeAssetType',
+  stepText: ' ',
+  isLoading: false,
+  type: 'submit',
+  priceType: 'fixed',
+  algorithmPriceType: 'free',
+  isBalanceSufficient: true,
+  isConsumable: true,
+  consumableFeedback: 'consumableFeedback',
+  isAlgorithmConsumable: true,
+  hasProviderFee: false,
+  retry: false
+}
+
+describe('Asset/AssetActions/ButtonBuy', () => {
+  //  TESTS FOR LOADING
+  it('Renders Buy button without crashing', async () => {
+    render(<ButtonBuy {...downloadProps} isLoading />)
+    const button = screen.getByText('TEST')
+    expect(button).toContainHTML('<Loader')
+  })
+
+  //  TESTS FOR DOWNLOAD
+  it('Renders Buy button without crashing', async () => {
+    render(<ButtonBuy {...downloadProps} />)
+    const button = screen.getByText('Buy for 1 day')
+    expect(button).toContainHTML('<button')
+  })
+
+  it('Renders Buy button without crashing when hasPreviousOrder=true', async () => {
+    downloadProps.hasPreviousOrder = true
```

Comment:
Object reassignments here and in all the following tests lead to modifying the global object. You reassign e.g. here `downloadProps.hasPreviousOrder` so the next test running after it will have `downloadProps.hasPreviousOrder = true`.

You have to put the overwrite props after the object spread on the component as suggested before, like so:

```
<ButtonBuy {...downloadProps} hasPreviousOrder />
```

This needs to be changed for each and every test block in here. Another thing is that none of the tests should be async, as you are not `await` for anything, so the `async` can be removed from all tests too.

Reply:
And despite the many tests in here and reporting 100% coverage, there are still large parts of the codebase untested. Because some messages are not triggered by any test:

<img width="758" alt="Screenshot 2022-11-10 at 10 55 40" src="https://user-images.githubusercontent.com/90316/201073040-b86df074-6736-4333-b1ed-ec69d9f2f22f.png">

From what I'm seeing, another good test would be everything around the disabled button. Like, line 100 deals with `isBalanceSufficient` so testing that button stays disabled/gets enabled when we want it to would cover that and probably more